File: src/manager.ts
import Anthropic from "@anthropic-ai/sdk";
import { Agent } from "./agent.js";
import { Dispatcher } from "./dispatcher.js";
import { type PeerApi, type SkillEnv } from "./skillkit.js";
import type { Store } from "./store.js";
import type { AppConfig, AgentStatus, BatchResult, BotSpec, CreateResult, SpawnResult } from "./types.js";

/** Owns the dispatcher and every worker; handles creation, ownership, and reuse. */
export class BotManager {
  private readonly agents = new Map<string, Agent>();
  private readonly dispatcher: Dispatcher;
  private readonly env: SkillEnv;
  private readonly client: Anthropic; // one client shared by the whole fleet
  private nextNumber: number;
  private maxPerUser: number;

  constructor(private readonly config: AppConfig, private readonly store: Store) {
    this.loadSettings(); // durable settings override the env-var seed
    this.maxPerUser = config.maxPerUser;
    this.client = new Anthropic({ apiKey: config.llm.apiKey });
    const peers: PeerApi = {
      position: (name) => this.agents.get(name)?.position() ?? null,
      online: (name) => !!this.agents.get(name)?.isOnline(),
      send: (to, from, message) => {
        const a = this.agents.get(to);
        if (!a || !a.isOnline()) return false;
        a.inject(`AGENT ${from}: ${message}`);
        return true;
      },
      summon: (count, goal, owner) => this.createNew(count, goal, owner),
    };
    this.env = { memory: store, peers, routines: store };
    this.dispatcher = new Dispatcher(config.dispatcherName, config.mc, config.commandAllowlist, {
      createNew: (count, goal, owner) => this.createNew(count, goal, owner),
      assignExisting: (nums, goal, owner) => this.assignExisting(nums, goal, owner),
      release: (nums, owner) => this.release(nums, owner),
      claim: (nums, owner) => this.claim(nums, owner),
      give: (nums, owner, target) => this.give(nums, owner, target),
    });
    for (const spec of config.bots) this.create(spec, null);
    this.nextNumber = config.bots.length + 1;
    this.restoreOwnership();
  }

  private loadSettings(): void {
    const { config, store } = this;
    const host = store.getSetting("mcHost");
    const port = store.getSetting("mcPort");
    const login = store.getSetting("loginMessage");
    const cap = store.getSetting("maxPerUser");
    if (host) config.mc.host = host;
    if (port) config.mc.port = Number(port);
    if (login != null) config.mc.loginMessage = login;
    if (cap != null) config.maxPerUser = Number(cap);
  }

  /** Recreate persisted owned numbers as offline placeholders so ownership survives restarts. */
  private restoreOwnership(): void {
    for (const { username, owner } of this.store.allAgents()) {
      const existing = this.agents.get(username);
      if (existing) { existing.owner = owner; continue; }
      this.create({ username }, owner).markOffline();
      const n = Number(username.replace(/^agent_/, ""));
      if (Number.isFinite(n)) this.nextNumber = Math.max(this.nextNumber, n + 1);
    }
  }

  private create(spec: BotSpec, owner: string | null): Agent {
    const agent = new Agent(spec, this.config.mc, this.config.llm, owner, this.env, this.client);
    this.agents.set(spec.username, agent);
    return agent;
  }

  /** Set an agent's owner and persist it (ownership is written on any change). */
  private setOwner(agent: Agent, name: string, owner: string | null): void {
    agent.owner = owner;
    this.store.setOwner(name, owner);
  }

  startAll(): void {
    this.dispatcher.start();
    this.dispatcher.enableRecycle(this.config.dispatcherRecycleMs);
    for (const agent of this.agents.values()) agent.start();
  }

  private onlineCount(): number {
    return [...this.agents.values()].filter((a) => a.isOnline()).length;
  }

  private ownedOnlineCount(owner: string): number {
    return [...this.agents.values()].filter((a) => a.owner === owner && a.isOnline()).length;
  }

  /** Per-user cap applies to real players, not the admin HTTP channel. */
  private overUserCap(owner: string | null): boolean {
    return owner !== null && owner !== "api" && this.maxPerUser > 0 && this.ownedOnlineCount(owner) >= this.maxPerUser;
  }

  /** Live settings shown/edited in the dashboard. */
  getSettings(): { maxBots: number; maxPerUser: number; mcHost: string; mcPort: number; loginMessage: string } {
    return {
      maxBots: this.config.maxBots,
      maxPerUser: this.maxPerUser,
      mcHost: this.config.mc.host,
      mcPort: this.config.mc.port,
      loginMessage: this.config.mc.loginMessage,
    };
  }

  /** Apply a live settings patch, persisting each change. Host/port/login changes reconnect the fleet. */
  updateSettings(patch: { maxPerUser?: number; mcHost?: string; mcPort?: number; loginMessage?: string }): void {
    if (patch.maxPerUser != null) {
      this.maxPerUser = Math.max(0, Math.floor(patch.maxPerUser));
      this.store.setSetting("maxPerUser", String(this.maxPerUser));
    }
    let reconnect = false;
    if (patch.mcHost != null && patch.mcHost !== this.config.mc.host) { this.config.mc.host = patch.mcHost; this.store.setSetting("mcHost", patch.mcHost); reconnect = true; }
    if (patch.mcPort != null && patch.mcPort !== this.config.mc.port) { this.config.mc.port = patch.mcPort; this.store.setSetting("mcPort", String(patch.mcPort)); reconnect = true; }
    if (patch.loginMessage != null && patch.loginMessage !== this.config.mc.loginMessage) { this.config.mc.loginMessage = patch.loginMessage; this.store.setSetting("loginMessage", patch.loginMessage); reconnect = true; }
    if (reconnect) {
      this.dispatcher.reconnect();
      for (const a of this.agents.values()) a.reconnect();
    }
  }

  private spawn(goal: string | undefined, owner: string | null): SpawnResult {
    if (this.overUserCap(owner)) return { ok: false, reason: "user_limit" };
    if (this.onlineCount() >= this.config.maxBots) return { ok: false, reason: "at_capacity" };
    let username = `agent_${this.nextNumber}`;
    while (this.agents.has(username)) username = `agent_${++this.nextNumber}`;
    this.create({ username, goal }, owner).start();
    this.store.setOwner(username, owner);
    this.nextNumber++;
    return { ok: true, username };
  }

  /** `new [n] <task>` — n fresh workers on one goal, owned by the caller. */
  createNew(count: number, goal: string, owner: string | null): CreateResult {
    const created: string[] = [];
    let rejected = 0;
    let reason: CreateResult["reason"];
    for (let i = 0; i < count; i++) {
      const r = this.spawn(goal, owner);
      if (r.ok && r.username) created.push(r.username);
      else {
        rejected++;
        reason = r.reason;
      }
    }
    return { created, rejected, reason };
  }

  /** `x[, y] <task>` — retask existing workers the caller owns. */
  assignExisting(numbers: number[], goal: string, owner: string): BatchResult {
    const done: string[] = [];
    const skipped: BatchResult["skipped"] = [];
    for (const name of numbers.map((n) => `agent_${n}`)) {
      const a = this.agents.get(name);
      if (!a) skipped.push({ name, reason: "unknown" });
      else if (a.owner !== owner) skipped.push({ name, reason: "not_owner" });
      else if (!a.isOnline() && this.overUserCap(owner)) skipped.push({ name, reason: "user_limit" });
      else if (!a.isOnline() && this.onlineCount() >= this.config.maxBots) skipped.push({ name, reason: "at_capacity" });
      else if (!a.assign(goal)) skipped.push({ name, reason: "busy" });
      else done.push(name);
    }
    return { done, skipped };
  }

  /** `release x[, y]` — the owner relinquishes ownership (becomes claimable). */
  release(numbers: number[], owner: string): BatchResult {
    const done: string[] = [];
    const skipped: BatchResult["skipped"] = [];
    for (const name of numbers.map((n) => `agent_${n}`)) {
      const a = this.agents.get(name);
      if (!a) skipped.push({ name, reason: "unknown" });
      else if (a.owner !== owner) skipped.push({ name, reason: "not_owner" });
      else {
        this.setOwner(a, name, null);
        done.push(name);
      }
    }
    return { done, skipped };
  }

  /** `claim x[, y]` — take an unowned number (creating it offline if new). */
  claim(numbers: number[], owner: string): BatchResult {
    const done: string[] = [];
    const skipped: BatchResult["skipped"] = [];
    for (const n of numbers) {
      const name = `agent_${n}`;
      const a = this.agents.get(name);
      if (!a) {
        this.create({ username: name }, owner).markOffline();
        this.store.setOwner(name, owner);
        this.nextNumber = Math.max(this.nextNumber, n + 1);
        done.push(name);
      } else if (a.owner === null || a.owner === owner) {
        this.setOwner(a, name, owner);
        done.push(name);
      } else skipped.push({ name, reason: "owned_by_other" });
    }
    return { done, skipped };
  }

  /** `give x[, y] <player>` — transfer ownership of your workers to another player. */
  give(numbers: number[], owner: string, target: string): BatchResult {
    const done: string[] = [];
    const skipped: BatchResult["skipped"] = [];
    for (const name of numbers.map((n) => `agent_${n}`)) {
      const a = this.agents.get(name);
      if (!a) skipped.push({ name, reason: "unknown" });
      else if (a.owner !== owner) skipped.push({ name, reason: "not_owner" });
      else {
        this.setOwner(a, name, target);
        done.push(name);
      }
    }
    return { done, skipped };
  }

  get(username: string): Agent | undefined {
    return this.agents.get(username);
  }

  list(): AgentStatus[] {
    return [...this.agents.values()].map((a) => a.status());
  }

  dispatcherStatus() {
    return this.dispatcher.status();
  }
}

File: src/manager.rs
//! Fleet manager (port of manager.ts). Owns the dispatcher + every worker and exposes the
//! facade api.rs consumes. Shared as `Arc<BotManager>`; all methods take `&self`.
use std::collections::HashMap;
use std::path::Path;
use std::sync::{Arc, OnceLock, Weak};

use anyhow::Result;
use parking_lot::Mutex;

use crate::agent::Agent; // TODO(verify): concurrent module; methods assumed &self w/ interior mutability
use crate::config::{normalize_model, MODELS};
use crate::dispatcher::{DispatchHandlers, Dispatcher}; // TODO(verify): concurrent module; trait shape assumed
use crate::library::{FileRoutineStore, FileRuleStore};
use crate::skill::{Memory, OwnerLookup, PeerApi, RoutineStore, RuleStore};
use crate::store::Store;
use crate::types::{
    AgentStatus, AppConfig, AssignOutcome, BatchResult, BotSpec, CreateResult, DispatcherStatus,
    LlmConfig, McConfig, Pos, RejectReason, Settings, SettingsPatch, Skipped,
};

/// Planner keys shared by the fleet. // TODO(verify): agent.rs consumes this shape.
#[derive(Clone)]
pub struct PlannerKeys {
    pub anthropic: String,
    pub openai: String,
}

/// The skill environment handed to each worker (mirrors TS `SkillEnv`).
/// TODO(verify): agent.rs constructs its SkillContext from these Arcs.
#[derive(Clone)]
pub struct SkillEnv {
    pub memory: Arc<dyn Memory>,
    pub peers: Arc<dyn PeerApi>,
    pub routines: Arc<dyn RoutineStore>,
    pub rules: Arc<dyn RuleStore>,
}

/// Op-teleport callback (dispatcher brings a worker to a player). TODO(verify): agent.rs param type.
pub type TeleportFn = Arc<dyn Fn(&str, &str) + Send + Sync>;

/// Per-spawn outcome (TS `SpawnResult`).
enum SpawnResult {
    Ok(String),
    Rejected(RejectReason),
}

/// Fleet-mutable state behind one lock (single-threaded in TS; interior-mutable here).
struct Fleet {
    agents: HashMap<String, Arc<Agent>>,
    next_number: usize,
    max_per_user: usize,
}

/// All shared manager state. Impls `PeerApi` + `DispatchHandlers`; wrapped in an Arc cycle
/// (env.peers / dispatcher handlers point back here) that lives for the process.
struct Inner {
    max_bots: usize,
    bots: Vec<BotSpec>,
    // live, shared by reference with every worker so settings edits reach them on reconnect.
    mc: Arc<Mutex<McConfig>>,
    llm: Arc<Mutex<LlmConfig>>,
    keys: PlannerKeys,
    store: Arc<Store>,
    routines: Arc<FileRoutineStore>,
    rules: Arc<FileRuleStore>,
    state: Mutex<Fleet>,
    dispatcher: OnceLock<Arc<Dispatcher>>,
    weak: OnceLock<Weak<Inner>>,
}

fn online_count(agents: &HashMap<String, Arc<Agent>>) -> usize {
    agents.values().filter(|a| a.is_online()).count()
}

fn owned_online_count(agents: &HashMap<String, Arc<Agent>>, owner: &str) -> usize {
    agents
        .values()
        .filter(|a| a.owner().as_deref() == Some(owner) && a.is_online())
        .count()
}

/// `agent_N` → N. Empty suffix reads as 0 (TS `Number("")`), non-numeric skips.
fn parse_agent_number(name: &str) -> Option<usize> {
    let rest = name.strip_prefix("agent_")?;
    if rest.is_empty() {
        Some(0)
    } else {
        rest.parse().ok()
    }
}

/// Durable settings override the env-var seed (TS `loadSettings`).
fn load_settings(config: &mut AppConfig, store: &Store) {
    if let Some(h) = store.get_setting("mcHost").filter(|s| !s.is_empty()) {
        config.mc.host = h;
    }
    if let Some(p) = store.get_setting("mcPort").filter(|s| !s.is_empty()) {
        if let Ok(n) = p.parse() {
            config.mc.port = n;
        }
    }
    if let Some(l) = store.get_setting("loginMessage") {
        config.mc.login_message = l;
    }
    if let Some(c) = store.get_setting("maxPerUser") {
        if let Ok(n) = c.parse() {
            config.max_per_user = n;
        }
    }
    if let Some(m) = store.get_setting("model").filter(|s| !s.is_empty()) {
        if let Ok(m) = normalize_model(&m) {
            config.llm.model = m;
        }
    }
    if let Some(s) = store.get_setting("maxSteps") {
        if let Ok(n) = s.parse() {
            config.llm.max_steps = n;
        }
    }
}

impl Inner {
    fn arc(&self) -> Arc<Inner> {
        self.weak.get().expect("weak self set").upgrade().expect("inner alive")
    }

    fn dispatcher(&self) -> &Arc<Dispatcher> {
        self.dispatcher.get().expect("dispatcher initialized")
    }

    /// Per-user cap applies to real players, not the admin HTTP channel.
    fn over_user_cap(&self, fleet: &Fleet, owner: Option<&str>) -> bool {
        match owner {
            Some(o) if o != "api" => {
                fleet.max_per_user > 0 && owned_online_count(&fleet.agents, o) >= fleet.max_per_user
            }
            _ => false,
        }
    }

    /// Construct a worker (shared live config + env) and register it. Does not start it.
    fn create(self: &Arc<Self>, fleet: &mut Fleet, spec: BotSpec, owner: Option<String>) -> Arc<Agent> {
        let username = spec.username.clone();
        let env = SkillEnv {
            memory: self.store.clone(),
            peers: Arc::clone(self) as Arc<dyn PeerApi>,
            routines: self.routines.clone(),
            rules: self.rules.clone(),
        };
        let disp = self.dispatcher().clone();
        let teleport: TeleportFn = Arc::new(move |name: &str, target: &str| disp.teleport(name, target));
        // TODO(verify): Agent::new signature/return type (assumed value, wrapped in Arc here).
        let agent = Arc::new(Agent::new(
            spec,
            self.mc.clone(),
            self.llm.clone(),
            owner,
            env,
            self.keys.clone(),
            teleport,
        ));
        fleet.agents.insert(username, agent.clone());
        agent
    }

    /// Set an agent's owner and persist it (ownership is written on any change).
    fn set_owner_persist(&self, agent: &Arc<Agent>, name: &str, owner: Option<&str>) {
        agent.set_owner(owner.map(str::to_string)); // TODO(verify): Agent::set_owner
        self.store.set_owner(name, owner);
    }

    fn spawn(self: &Arc<Self>, fleet: &mut Fleet, goal: Option<&str>, owner: Option<&str>) -> SpawnResult {
        if self.over_user_cap(fleet, owner) {
            return SpawnResult::Rejected(RejectReason::UserLimit);
        }
        if online_count(&fleet.agents) >= self.max_bots {
            return SpawnResult::Rejected(RejectReason::AtCapacity);
        }
        let mut username = format!("agent_{}", fleet.next_number);
        while fleet.agents.contains_key(&username) {
            fleet.next_number += 1;
            username = format!("agent_{}", fleet.next_number);
        }
        let spec = BotSpec { username: username.clone(), goal: goal.map(str::to_string), model: None };
        let agent = self.create(fleet, spec, owner.map(str::to_string));
        agent.start();
        self.store.set_owner(&username, owner);
        fleet.next_number += 1;
        SpawnResult::Ok(username)
    }

    /// `new [n] <task>` — n fresh workers on one goal, owned by the caller.
    fn create_new(self: &Arc<Self>, count: usize, goal: &str, owner: Option<&str>) -> CreateResult {
        let mut fleet = self.state.lock();
        let mut created = Vec::new();
        let mut rejected = 0usize;
        let mut reason = None;
        for _ in 0..count {
            match self.spawn(&mut fleet, Some(goal), owner) {
                SpawnResult::Ok(u) => created.push(u),
                SpawnResult::Rejected(r) => {
                    rejected += 1;
                    reason = Some(r);
                }
            }
        }
        CreateResult { created, rejected, reason }
    }

    /// `x[, y] <task>` — retask existing workers the caller owns.
    fn assign_existing(self: &Arc<Self>, numbers: &[u32], goal: &str, owner: &str) -> BatchResult {
        let mut done = Vec::new();
        let mut skipped = Vec::new();
        let fleet = self.state.lock();
        for n in numbers {
            let name = format!("agent_{n}");
            let a = fleet.agents.get(&name).cloned();
            match a {
                None => skipped.push(Skipped { name, reason: "unknown".into() }),
                Some(a) if a.owner().as_deref() != Some(owner) => {
                    skipped.push(Skipped { name, reason: "not_owner".into() })
                }
                Some(a) if !a.is_online() && self.over_user_cap(&fleet, Some(owner)) => {
                    skipped.push(Skipped { name, reason: "user_limit".into() })
                }
                Some(a) if !a.is_online() && online_count(&fleet.agents) >= self.max_bots => {
                    skipped.push(Skipped { name, reason: "at_capacity".into() })
                }
                Some(a) if !a.assign(goal) => skipped.push(Skipped { name, reason: "busy".into() }),
                Some(_) => done.push(name),
            }
        }
        BatchResult { done, skipped }
    }

    /// `free x[, y]` — the owner relinquishes ownership (becomes claimable).
    fn free(self: &Arc<Self>, numbers: &[u32], owner: &str) -> BatchResult {
        let mut done = Vec::new();
        let mut skipped = Vec::new();
        let fleet = self.state.lock();
        for n in numbers {
            let name = format!("agent_{n}");
            match fleet.agents.get(&name).cloned() {
                None => skipped.push(Skipped { name, reason: "unknown".into() }),
                Some(a) if a.owner().as_deref() != Some(owner) => {
                    skipped.push(Skipped { name, reason: "not_owner".into() })
                }
                Some(a) => {
                    self.set_owner_persist(&a, &name, None);
                    done.push(name);
                }
            }
        }
        BatchResult { done, skipped }
    }

    /// `claim x[, y]` — take an unowned number (creating it offline if new).
    fn claim(self: &Arc<Self>, numbers: &[u32], owner: &str) -> BatchResult {
        let mut done = Vec::new();
        let mut skipped = Vec::new();
        let mut fleet = self.state.lock();
        for n in numbers {
            let name = format!("agent_{n}");
            match fleet.agents.get(&name).cloned() {
                None => {
                    let spec = BotSpec { username: name.clone(), goal: None, model: None };
                    let agent = self.create(&mut fleet, spec, Some(owner.to_string()));
                    agent.mark_offline();
                    self.store.set_owner(&name, Some(owner));
                    fleet.next_number = fleet.next_number.max(*n as usize + 1);
                    done.push(name);
                }
                Some(a) => {
                    let o = a.owner();
                    if o.is_none() || o.as_deref() == Some(owner) {
                        self.set_owner_persist(&a, &name, Some(owner));
                        done.push(name);
                    } else {
                        skipped.push(Skipped { name, reason: "owned_by_other".into() });
                    }
                }
            }
        }
        BatchResult { done, skipped }
    }

    /// `quit x[, y]` — immediately disconnect workers the caller owns, even mid-task.
    fn quit(self: &Arc<Self>, numbers: &[u32], owner: &str) -> BatchResult {
        let mut done = Vec::new();
        let mut skipped = Vec::new();
        let fleet = self.state.lock();
        for n in numbers {
            let name = format!("agent_{n}");
            match fleet.agents.get(&name).cloned() {
                None => skipped.push(Skipped { name, reason: "unknown".into() }),
                Some(a) if a.owner().as_deref() != Some(owner) => {
                    skipped.push(Skipped { name, reason: "not_owner".into() })
                }
                Some(a) => {
                    a.stop();
                    done.push(name);
                }
            }
        }
        BatchResult { done, skipped }
    }

    /// `give x[, y] <player>` — transfer ownership of your workers to another player.
    fn give(self: &Arc<Self>, numbers: &[u32], owner: &str, target: &str) -> BatchResult {
        let mut done = Vec::new();
        let mut skipped = Vec::new();
        let fleet = self.state.lock();
        for n in numbers {
            let name = format!("agent_{n}");
            match fleet.agents.get(&name).cloned() {
                None => skipped.push(Skipped { name, reason: "unknown".into() }),
                Some(a) if a.owner().as_deref() != Some(owner) => {
                    skipped.push(Skipped { name, reason: "not_owner".into() })
                }
                Some(a) => {
                    self.set_owner_persist(&a, &name, Some(target));
                    done.push(name);
                }
            }
        }
        BatchResult { done, skipped }
    }

    /// Recreate persisted owned numbers as offline placeholders so ownership survives restarts.
    fn restore_ownership(self: &Arc<Self>) {
        let rows = self.store.all_agents();
        let mut fleet = self.state.lock();
        for (username, owner) in rows {
            if let Some(existing) = fleet.agents.get(&username) {
                existing.set_owner(owner); // TODO(verify): Agent::set_owner (no persist; already from store)
                continue;
            }
            let spec = BotSpec { username: username.clone(), goal: None, model: None };
            let agent = self.create(&mut fleet, spec, owner);
            agent.mark_offline();
            if let Some(n) = parse_agent_number(&username) {
                fleet.next_number = fleet.next_number.max(n + 1);
            }
        }
    }

    fn wipe_agents(&self) -> usize {
        let mut fleet = self.state.lock();
        let removed = fleet.agents.len();
        for a in fleet.agents.values() {
            a.stop();
        }
        fleet.agents.clear();
        self.store.wipe_agent_data();
        fleet.next_number = 1;
        removed
    }
}

impl PeerApi for Inner {
    fn position(&self, name: &str) -> Option<Pos> {
        self.state.lock().agents.get(name).and_then(|a| a.position())
    }
    fn online(&self, name: &str) -> bool {
        self.state.lock().agents.get(name).map_or(false, |a| a.is_online())
    }
    fn send(&self, to: &str, from: &str, message: &str) -> bool {
        let a = self.state.lock().agents.get(to).cloned();
        match a {
            Some(a) if a.is_online() => {
                a.inject(&format!("AGENT {from}: {message}"));
                true
            }
            _ => false,
        }
    }
    fn owner_of(&self, name: &str) -> OwnerLookup {
        self.state.lock().agents.get(name).map(|a| a.owner())
    }
    fn teammates(&self, owner: Option<&str>) -> Vec<String> {
        match owner {
            None => Vec::new(),
            Some(o) => self
                .state
                .lock()
                .agents
                .iter()
                .filter(|(_, a)| a.owner().as_deref() == Some(o) && a.is_online())
                .map(|(n, _)| n.clone())
                .collect(),
        }
    }
    fn summon(&self, count: usize, goal: &str, owner: Option<&str>) -> CreateResult {
        self.arc().create_new(count, goal, owner)
    }
}

// Dispatcher command grammar handlers. Owner is always the commanding player here.
impl DispatchHandlers for Inner {
    fn create_new(&self, count: usize, goal: &str, owner: &str) -> CreateResult {
        self.arc().create_new(count, goal, Some(owner))
    }
    fn assign_existing(&self, numbers: &[u32], goal: &str, owner: &str) -> BatchResult {
        self.arc().assign_existing(numbers, goal, owner)
    }
    fn free(&self, numbers: &[u32], owner: &str) -> BatchResult {
        self.arc().free(numbers, owner)
    }
    fn claim(&self, numbers: &[u32], owner: &str) -> BatchResult {
        self.arc().claim(numbers, owner)
    }
    fn quit(&self, numbers: &[u32], owner: &str) -> BatchResult {
        self.arc().quit(numbers, owner)
    }
    fn give(&self, numbers: &[u32], owner: &str, target: &str) -> BatchResult {
        self.arc().give(numbers, owner, target)
    }
}

/// Owns the dispatcher and every worker; handles creation, ownership, and reuse.
pub struct BotManager {
    inner: Arc<Inner>,
}

impl BotManager {
    pub fn new(config: AppConfig, store: Arc<Store>) -> BotManager {
        let mut config = config;
        load_settings(&mut config, &store); // durable settings override the env-var seed
        let keys = PlannerKeys { anthropic: config.llm.api_key.clone(), openai: config.llm.openai_api_key.clone() };
        // One shared library for the whole fleet: routines/ and settings/ as JSON files (not the DB).
        let lib = Path::new(&config.library_dir);
        let routines = Arc::new(FileRoutineStore::new(lib.join("routines")));
        let rules = Arc::new(FileRuleStore::new(lib.join("settings")));
        let inner = Arc::new(Inner {
            max_bots: config.max_bots,
            bots: config.bots.clone(),
            mc: Arc::new(Mutex::new(config.mc.clone())),
            llm: Arc::new(Mutex::new(config.llm.clone())),
            keys,
            store,
            routines,
            rules,
            state: Mutex::new(Fleet { agents: HashMap::new(), next_number: 0, max_per_user: config.max_per_user }),
            dispatcher: OnceLock::new(),
            weak: OnceLock::new(),
        });
        inner.weak.set(Arc::downgrade(&inner)).ok();
        // Dispatcher holds the handlers (this manager); teleport callbacks reference it.
        // TODO(verify): Dispatcher::new signature (shared live mc; handlers as Arc<dyn DispatchHandlers>).
        let dispatcher = Arc::new(Dispatcher::new(
            config.dispatcher_name.clone(),
            inner.mc.clone(),
            config.command_allowlist.clone(),
            inner.clone() as Arc<dyn DispatchHandlers>,
        ));
        inner.dispatcher.set(dispatcher).ok();
        {
            let mut fleet = inner.state.lock();
            for spec in inner.bots.iter() {
                inner.create(&mut fleet, spec.clone(), None);
            }
            fleet.next_number = inner.bots.len() + 1;
        }
        inner.restore_ownership();
        BotManager { inner }
    }

    pub fn start_all(&self) {
        let inner = &self.inner;
        inner.dispatcher().start();
        let agents: Vec<Arc<Agent>> = inner.state.lock().agents.values().cloned().collect();
        for a in agents {
            a.start();
        }
    }

    // ---- facade (api.rs consumes exactly these) ----

    pub fn list(&self) -> Vec<AgentStatus> {
        self.inner.state.lock().agents.values().map(|a| a.status()).collect()
    }

    pub fn status(&self, name: &str) -> Option<AgentStatus> {
        self.inner.state.lock().agents.get(name).map(|a| a.status())
    }

    pub fn dispatcher_status(&self) -> DispatcherStatus {
        self.inner.dispatcher().status()
    }

    /// Live settings shown/edited in the dashboard.
    pub fn get_settings(&self) -> Settings {
        let (mc_host, mc_port, login_message) = {
            let mc = self.inner.mc.lock();
            (mc.host.clone(), mc.port, mc.login_message.clone())
        };
        let (model, max_steps) = {
            let llm = self.inner.llm.lock();
            (llm.model.clone(), llm.max_steps)
        };
        Settings {
            max_bots: self.inner.max_bots,
            max_per_user: self.inner.state.lock().max_per_user,
            mc_host,
            mc_port,
            login_message,
            model,
            models: MODELS.iter().map(|s| s.to_string()).collect(),
            max_steps,
        }
    }

    /// Apply a live settings patch, persisting each change. Host/port/login reconnect the fleet;
    /// model/step-budget/cap apply to each worker's next (or in-flight) task.
    pub fn update_settings(&self, patch: SettingsPatch) -> Result<()> {
        let inner = &self.inner;
        if let Some(mpu) = patch.max_per_user {
            inner.state.lock().max_per_user = mpu;
            inner.store.set_setting("maxPerUser", &mpu.to_string());
        }
        if let Some(model) = patch.model.as_deref() {
            let model = normalize_model(model)?; // Err on an unknown model
            inner.llm.lock().model = model.clone();
            inner.store.set_setting("model", &model);
        }
        if let Some(ms) = patch.max_steps {
            let ms = ms.clamp(1, 1000);
            inner.llm.lock().max_steps = ms;
            inner.store.set_setting("maxSteps", &ms.to_string());
        }
        let mut reconnect = false;
        {
            let mut mc = inner.mc.lock();
            if let Some(h) = patch.mc_host.as_deref() {
                if h != mc.host {
                    mc.host = h.to_string();
                    inner.store.set_setting("mcHost", h);
                    reconnect = true;
                }
            }
            if let Some(p) = patch.mc_port {
                if p != mc.port {
                    mc.port = p;
                    inner.store.set_setting("mcPort", &p.to_string());
                    reconnect = true;
                }
            }
            if let Some(l) = patch.login_message.as_deref() {
                if l != mc.login_message {
                    mc.login_message = l.to_string();
                    inner.store.set_setting("loginMessage", l);
                    reconnect = true;
                }
            }
        }
        if reconnect {
            inner.dispatcher().reconnect();
            let agents: Vec<Arc<Agent>> = inner.state.lock().agents.values().cloned().collect();
            for a in agents {
                a.reconnect();
            }
        }
        Ok(())
    }

    pub fn create_new(&self, count: usize, goal: &str, owner: Option<&str>) -> CreateResult {
        self.inner.create_new(count, goal, owner)
    }

    pub fn assign(&self, name: &str, goal: &str) -> AssignOutcome {
        let a = self.inner.state.lock().agents.get(name).cloned();
        match a {
            None => AssignOutcome::NotFound,
            Some(a) => {
                if a.assign(goal) {
                    AssignOutcome::Ok
                } else {
                    AssignOutcome::Busy
                }
            }
        }
    }

    pub fn chat(&self, name: &str, message: &str) -> bool {
        let a = self.inner.state.lock().agents.get(name).cloned();
        match a {
            Some(a) => {
                a.chat(message);
                true
            }
            None => false,
        }
    }

    pub fn stop(&self, name: &str) -> bool {
        let a = self.inner.state.lock().agents.get(name).cloned();
        match a {
            Some(a) => {
                a.stop();
                true
            }
            None => false,
        }
    }

    /// Dev reset: disconnect and forget every worker + its memory. Keeps live settings and the
    /// shared library. Dispatcher stays online.
    pub fn wipe_agents(&self) -> usize {
        self.inner.wipe_agents()
    }
}
